src/hooks/usePermit2Allowance.ts

```diff
@@ -0,0 +1,122 @@
+import { usePendingApproval } from 'hooks/transactions'
+import useInterval from 'hooks/useInterval'
+import { PermitSignature, usePermitAllowance, useUpdatePermitAllowance } from 'hooks/usePermitAllowance'
+import { useTokenAllowance, useUpdateTokenAllowance } from 'hooks/useTokenAllowance'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+import { ApprovalTransactionInfo } from 'state/transactions'
+
+enum ApprovalState {
+  PENDING,
+  SYNCING,
+  SYNCED,
+}
+
+export enum AllowanceState {
+  UNKNOWN,
+  REQUIRED,
+  ALLOWED,
+}
+
+export interface AllowanceRequired {
+  state: AllowanceState.REQUIRED
+  token: Token
+  isApprovalLoading: boolean
+  callback: () => Promise<ApprovalTransactionInfo | void>
+}
+
+export type Allowance =
+  | { state: AllowanceState.UNKNOWN }
+  | {
+      state: AllowanceState.ALLOWED
+      signature?: PermitSignature
+    }
+  | AllowanceRequired
+
+export default function usePermit2Allowance(amount?: CurrencyAmount<Token>, spender?: string): Allowance {
+  const { account } = useWeb3React()
+  const token = amount?.currency
+
+  const { tokenAllowance, isSyncing: isApprovalSyncing } = useTokenAllowance(token, account, PERMIT2_ADDRESS)
+  const updateTokenAllowance = useUpdateTokenAllowance(amount, PERMIT2_ADDRESS)
+  const isApproved = useMemo(() => {
+    if (!amount || !tokenAllowance) return false
+    return tokenAllowance.greaterThan(amount) || tokenAllowance.equalTo(amount)
+  }, [amount, tokenAllowance])
+
+  // Marks approval as loading from the time it is submitted (pending), until it has confirmed and another block synced.
+  // This avoids re-prompting the user for an already-submitted but not-yet-observed approval, by marking it loading
+  // until it has been re-observed. It wll sync immediately, because confirmation fast-forwards the block number.
+  const [approvalState, setApprovalState] = useState(ApprovalState.SYNCED)
+  const isApprovalLoading = approvalState !== ApprovalState.SYNCED
+  const isApprovalPending = Boolean(usePendingApproval(token, PERMIT2_ADDRESS))
+  useEffect(() => {
+    if (isApprovalPending) {
+      setApprovalState(ApprovalState.PENDING)
+    } else {
+      setApprovalState((state) => {
+        if (state === ApprovalState.PENDING && isApprovalSyncing) {
+          return ApprovalState.SYNCING
+        } else if (state === ApprovalState.SYNCING && !isApprovalSyncing) {
+          return ApprovalState.SYNCED
+        }
+        return state
+      })
+    }
+  }, [isApprovalPending, isApprovalSyncing])
+
+  // Signature and PermitAllowance will expire, so they should be rechecked at an interval.
+  const [now, setNow] = useState(Date.now())
+  // Calculate now such that the signature will still be valid for the submitting block.
+  useInterval(() => setNow((Date.now() - STANDARD_L1_BLOCK_TIME) / 1000), STANDARD_L1_BLOCK_TIME, true)
+
+  const [signature, setSignature] = useState<PermitSignature>()
+  const isSigned = useMemo(() => {
```

Comment:
this does seem worth testing as a unit test fwiw, it's pretty critical

Reply:
yes. this is flag-guarded, and tests will be coming in subsequent PRs.